### PR TITLE
[SPARK-34035][SQL] Refactor ScriptTransformation to remove input parameter and replace it by child.output

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Refactor ScriptTransformation to remove input parameter and replace it by child.output

### Why are the changes needed?
refactor code

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existed UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
FYI @maropu

**Comment by maropu** [comment]
Could you fix the build failure?

**Comment by AngersZhuuuu** [comment]
> Could you fix the build failure?

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -772,9 +772,6 @@ object ColumnPruning extends Rule[LogicalPlan] {
       f.copy(child = prunedChild(child, f.references))
     case e @ Expand(_, _, child) if !child.outputSet.subsetOf(e.references) =>
       e.copy(child = prunedChild(child, e.references))
-    case s @ ScriptTransformation(_, _, _, child, _)
-        if !child.outputSet.subsetOf(s.references) =>
-      s.copy(child = prunedChild(child, s.references))
```
Comment (AngersZhuuuu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:777]: Remove this since we don't need column pruning here

**Comment by AngersZhuuuu** [comment]
Gentle ping @maropu @cloud-fan This pr pass jenkins test now

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ScriptTransformation.scala [hunk 2]
```diff
@@ -17,24 +17,22 @@
...
  */
 case class ScriptTransformation(
-    input: Seq[Expression],
     script: String,
     output: Seq[Attribute],
     child: LogicalPlan,
     ioschema: ScriptInputOutputSchema) extends UnaryNode {
   @transient
-  override lazy val references: AttributeSet = AttributeSet(input.flatMap(_.references))
+  override lazy val references: AttributeSet = AttributeSet(child.output.flatMap(_.references))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ScriptTransformation.scala:35]: nit: `children.output`, as it's already `Seq[Attribute]`

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ScriptTransformation.scala:35]: > nit: `child.output.flatMap(_.references)` -> `children.output`, as it's already `Seq[Attribute]`

DOne

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:777]: Why don't we need the pruning?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:777]: > Why don't we need the pruning?

it pruning input to child, child's output always transform's input, so the column pruning for script transform is not useful.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:777]: Ah, okay. I remembered that `ScriptTransformation` always has a projected list of exprs in it's child now.

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**